src/NHibernate/Loader/Loader.cs

```diff
@@ -1829,7 +1831,36 @@ protected IList List(ISessionImplementor session, QueryParameters queryParameter
 
 		internal bool IsCacheable(QueryParameters queryParameters)
 		{
-			return _factory.Settings.IsQueryCacheEnabled && queryParameters.Cacheable;
+			if (Factory.Settings.IsQueryCacheEnabled && queryParameters.Cacheable)
+			{
+				if (QuerySpaces?.Count > 0)
+				{
+					ISet<string> neverCachedEntities = new HashSet<string>();
+
+					foreach (var persister in _factory
+						.GetEntityPersisters(QuerySpaces)
```

Comment:
When we think about never cached entity, gets high concurrent dml so it is not a good practise to cache with second level cache with another entity.

But you are right we need to add SupportsQueryCache on IQueryable and find through persisters not from tables.
How to implement it where is the best place to put this logic? I am not satisfied current implementation too.

Reply:
> I think it should be implemented some other way

You are correct, as you mentioned with query spaces we unfortunately cannot get the exact persisters, so we would need to detect the persisters for each query api separately.

For hql we could:
- find the persisters inside `HqlSqlWalker` by checking the usages for `HqlSqlWalker.AddQuerySpaces` method
- pre-calculate whether the query is cacheable and store the result in `QueryTranslatorImpl`
- make `Loader.IsCacheable` method virtual and override it in `QueryLoader` where the check could be implemented

for criteria:
- the persisters can be retrieved by using the logic from `CriteriaQueryTranslator.GetQuerySpaces` (we would need to extend `ICriteriaInfoProvider` for returning the persister)
- pre-calculate whether the query is cacheable and store the result in `CriteriaQueryTranslator`
- override `Loader.IsCacheable` to check whether the query is cacheable by using the pre-calculated value

for custom sql:
- the logic for retrieving the persister is located in the `CustomLoader` constructor, but for spaces added by `AddSynchronizedQuerySpace` we would need to guess the persisters
- in the `CustomLoader` constructor we could pre-calculate whether the query is cacheable
- override `Loader.IsCacheable` to check whether the query is cacheable by using the pre-calculated value

Reply:
Is it better option returning SupportsCacheable in `HqlSqlWalker` and `ICriteriaInfoProvider` vs returning persisters/collection persisters in hql/linq/criteria cases. If you insist returning persisters I will change it. 
I just don't want to expose whole persisters/collection persisters interface.

> we would need to extend ICriteriaInfoProvider for returning the persister

I understand that we  need to add property in `ICriteriaInfoProvider` is it not breaking change?

Reply:
> Is it better option returning SupportsCacheable in HqlSqlWalker and ICriteriaInfoProvider

Unfortunately this is not enough as we need the persisters to report the actual used persisters, otherwise we have the same issue that @bahusoid described. I did some modifications so that `IsCacheable` method now gets the persisters instead of query spaces.

> I understand that we need to add property in ICriteriaInfoProvider is it not breaking change?

Every change to a public interface is a breaking change, by extending I meant to create a temporal interface similar to `SessionFactoryExtension`.

Reply:
Thank you all your effort @maca88, I am glad to have a change to create new feature.

Reply:
@maca88  can we publish this feature in the next minor version, do I need to make any other changes?

Reply:
Based on the conversation in #2735, we may need to add a setting that would log a warning instead of throwing an exception for scenarios where query composition is commom.

Reply:
> Query composition is common
Did you mean cached query contains both cacheable and non cacheable spaces, if setting value is false(default true?) then just add warning log instead of throwing exception.

If you give me setting name(query_cache_throw_never_cached), I could implement it but I think it could lead unexpected issues.

Reply:
> Did you mean cached query contains both cacheable and non cacheable spaces, if setting value is false(default true?) then just add warning log instead of throwing exception.

Yes. Here is an example:
```
var query = session.Query<CachableEntity>()
    .WithOptions(x => x.SetCacheable(true))
    .Where(o => o.Name == filter.Name);
if (filter.Age.HasValue) {
  query = query.Where(o => o.NeverCachedEntity.Age == filter.Age);
}
```

in the above case the query will be cached only when `Age` filter is not set.

> If you give me setting name(query_cache_throw_never_cached), I could implement it but I think it could lead unexpected issues.

To be honest I am not sure what setting name would be the best for such setting, maybe `query.throw_never_cached`. What kind of unexpected issues do you have in mind? I think is fine if someone would like to have the above query cacheable when possible.

Reply:
Allowing cacheable query with never cached entity, it may mis lead developer and could be runtime stale data.
As I see its optinal defaults true and developer knows may lead stale data then ok its good to have it.
Is there any difference filter or reference property(CachedEntity.NeverCachedEntity.Age==3)usage on your sample?

Reply:
If the query contains a never cached query space, it should not be cached. The logging and/or exception should indicate that caching was ignored.

Reply:
ok

Reply:
Hi @maca88 will u accept changes, is there anything waiting to fix?

Reply:
I will go through the PR once more this weekend and add my final review.